packages/block-library/src/columns/edit.js

```diff
@@ -159,4 +191,37 @@ export default withDispatch( ( dispatch, ownProps, registry ) => ( {
 
 		replaceInnerBlocks( clientId, innerBlocks, false );
 	},
+
+	updateColumnWidth( clientId, width ) {
+		const { updateBlockAttributes } = dispatch( 'core/block-editor' );
+		const { getBlockRootClientId, getBlocks } = registry.select( 'core/block-editor' );
+
+		// Constrain or expand siblings to account for gain or loss of
+		// total columns area.
+		const columns = getBlocks( getBlockRootClientId( clientId ) );
+		const adjacentColumns = getAdjacentBlocks( columns, clientId );
+
+		// The occupied width is calculated as the sum of the new width
+		// and the total width of blocks _not_ in the adjacent set.
+		const occupiedWidth = width + getTotalColumnsWidth(
+			difference( columns, [
+				find( columns, { clientId } ),
+				...adjacentColumns,
+			] )
+		);
+
+		// Compute _all_ next column widths, in case the updated column
+		// is in the middle of a set of columns which don't yet have
+		// any explicit widths assigned (include updates to those not
+		// part of the adjacent blocks).
+		const nextColumnWidths = {
+			...getColumnWidths( columns, columns.length ),
+			[ clientId ]: toWidthPrecision( width ),
+			...getRedistributedColumnWidths( adjacentColumns, 100 - occupiedWidth, columns.length ),
+		};
+
+		forEach( nextColumnWidths, ( nextColumnWidth, columnClientId ) => {
+			updateBlockAttributes( columnClientId, { width: nextColumnWidth } );
```

Comment:
It would be great to make `updateBlockAttributes` more flexible or introduce some other action for batch updates like this one. In the case when there are 6 columns, it might trigger re-render 6 times for every atomic block update. Regardless of performance considerations, I anticipate that with nested blocks starting to be more and more popular we need to start thinking about batching actions to make APIs simpler to use.

Reply:
Another issue here might be that we add a number of Undo levels for each change.

Reply:
It feels like we need to start working on undo transactions soon. This has came up in other PRs recently.

Reply:
> It feels like we need to start working on undo transactions soon. This has came up in other PRs recently.

Yeah, it's similar to #8119, but I wonder if there might be some simple near-term solution:

```js
wp.data.dispatch( 'core/editor' ).transact( () => {
	wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( /* ... */ );
	wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( /* ... */ );
} );
// `'START_TRANSACTION'`
// `'UPDATE_BLOCK_ATTRIBUTES'`
// `'UPDATE_BLOCK_ATTRIBUTES'`
// `'END_TRANSACTION'`
```

(or just `startTransaction` + `endTransaction` / `toggleTransaction` actions)

Probably worth a separate issue for tracking and discussion.